### DISCUSSION TITLE
Dumping Selenium

### DISCUSSION THREAD

**Discussion by barnabasharris** [h-a-graham/EAlidaR#48] [Ideas]
Hi Hugh,
Hope you're well. After mucking around with chrome versions for the umpteenth time I thought that you'd be interested to hear that I _think_ I've found away to do away with Selenium entirely. The below script uses JSON requests to emulate the user interaction. I developed it on my Mac (hence the Mac specific User-Agent parameters) but then pasted it onto my Ubuntu machine's R terminal and it ran first time, so here's hoping it's robust.

Be great to integrate it into the package at some point.

```
require(httr)
require(sf)
require(geojsonsf)
require(stringr)
require(jsonlite)

# define polygon
aoi <- st_point(c(418978.7,511490.4)) %>% 
  st_buffer(500,endCapStyle = 'SQUARE') %>% 
  st_bbox() %>% st_as_sfc() %>% st_sf %>% 
  st_set_crs(27700)
# convert to geojson and parse vertices
polVerts <- geojsonsf::sf_geojson(aoi) %>% 
  as.character() %>% 
  stringr::str_extract('\\[\\[(.*?)\\]\\]') %>% 
  paste0(.,']')

# make request with polygon
headers = c(
  `User-Agent` = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:103.0) Gecko/20100101 Firefox/103.0',
  `Accept` = '*/*',
  `Accept-Language` = 'en-US,en;q=0.5',
  `Accept-Encoding` = 'gzip, deflate, br',
  `Connection` = 'keep-alive',
  `Referer` = 'https://environment.data.gov.uk/DefraDataDownload/?Mode=survey',
  `Sec-Fetch-Dest` = 'empty',
  `Sec-Fetch-Mode` = 'cors',
  `Sec-Fetch-Site` = 'same-origin',
  `If-None-Match` = 'cad21b43'
)

# insert polygons into params
params = list(
  `f` = 'json',
  `SourceToken` = '',
  `OutputFormat` = '0',
  `RequestMode` = 'SURVEY',
  `AOI` = paste0('{"geometryType":"esriGeometryPolygon","features":[{"geometry":{"rings":',
                 polVerts,
                 ',"spatialReference":{"wkid":27700,"latestWkid":27700}}}],"sr":{"wkid":27700,"latestWkid":27700}}')
)

res <- 
  httr::GET(url = 
              'https://environment.data.gov.uk/arcgis/rest/services/gp/DataDownload/GPServer/DataDownload/submitJob',
            httr::add_headers(.headers=headers), query = params)

http_error(res)
jsonRespParsed <- content(res,as="parsed") 
jobId <- jsonRespParsed$jobId
jsonRespParsed

# submit queries until results
cookies = c(
  'AGS_ROLES' = '"419jqfa+uOZgYod4xPOQ8Q=="'
)

headers = c(
  `User-Agent` = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:103.0) Gecko/20100101 Firefox/103.0',
  `Accept` = '*/*',
  `Accept-Language` = 'en-US,en;q=0.5',
  `Accept-Encoding` = 'gzip, deflate, br',
  `Connection` = 'keep-alive',
  `Referer` = 'https://environment.data.gov.uk/DefraDataDownload/?Mode=survey',
  `Sec-Fetch-Dest` = 'empty',
  `Sec-Fetch-Mode` = 'cors',
  `Sec-Fetch-Site` = 'same-origin'
)

params = list(
  `f` = 'json',
  `dojo.preventCache` = '1662031406539'
)

res <- httr::GET(url = 
                   paste0('https://environment.data.gov.uk/arcgis/rest/services/gp/DataDownload/GPServer/DataDownload/jobs/',
                          jobId),
                 httr::add_headers(.headers=headers), 
                 query = params, httr::set_cookies(.cookies = cookies))
jsonRespParsed <- content(res,as="parsed") 

while (jsonRespParsed$jobStatus %in% c("esriJobSubmitted", 'esriJobExecuting')) {
  print(jsonRespParsed$jobStatus)
  Sys.sleep(2)
  res <- httr::GET(url = 
                     paste0('https://environment.data.gov.uk/arcgis/rest/services/gp/DataDownload/GPServer/DataDownload/jobs/',
                            jobId),
                   httr::add_headers(.headers=headers), 
                   query = params, httr::set_cookies(.cookies = cookies))
  jsonRespParsed <- content(res,as="parsed") 
}

print(jsonRespParsed$jobStatus)


# get results

headers = c(
  `User-Agent` = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:103.0) Gecko/20100101 Firefox/103.0',
  `Accept` = '*/*',
  `Accept-Language` = 'en-US,en;q=0.5',
  `Accept-Encoding` = 'gzip, deflate, br',
  `Connection` = 'keep-alive',
  `Referer` = 'https://environment.data.gov.uk/DefraDataDownload/?Mode=survey',
  `Sec-Fetch-Dest` = 'empty',
  `Sec-Fetch-Mode` = 'cors',
  `Sec-Fetch-Site` = 'same-origin'
)

params = list(
  `f` = 'json',
  `dojo.preventCache` = '1662033674668'
)

res <- httr::GET(url = 
                   paste0('https://environment.data.gov.uk/arcgis/rest/services/gp/DataDownload/GPServer/DataDownload/jobs/',
                          jobId,
                          '/results/OutputResult'), 
                 httr::add_headers(.headers=headers), query = params, httr::set_cookies(.cookies = cookies))
http_error(res)
jsonRespParsed<-content(res,as="parsed") 


res <- fromJSON(jsonRespParsed$value$url)
res


```

**Comment by h-a-graham** [comment]
Wow! Barney, this is just brilliant! You actually figured out how to work with the rest API?! So, This has to be incorporated - it is so much faster and simpler than Selenium. I have some questions and I definitely don't fully understnad everything that is going on here yet. But this is totally the way to go! 

Do you know if this will enable us to increase the request area for tiles? A quick play suggests that there are some limits but these might be much larger than for the portal? 

I think we can nicely slot this into the current approach of calculating tile intersections and then making individual tile OS tile requests... but maybe we can just ask for more?

It's been a while since I've looked at this package much but this has got me excited to do some long over due tidying/refactoring!

Thankyou!

**Reply by barnabasharris** [reply]
My pleasure. Once I'd started fiddling with the JSON stuff I couldn't stop until it was functional.

First, the method I used to devise the script (which could apply to most websites, I believe) was to browse to the portal and then turn on Firefox firebug inspector, go to the network tab and examine the calls taking place as I clicked the 'Get Tiles' button etc.

<img width="1678" alt="image" src="https://user-images.githubusercontent.com/71077322/187958541-b56888c0-d747-4353-9a44-9a058c99f11c.png">

Then I copied the GET requests values as cURL commands (see picture) and used [this website](https://curlconverter.com/r) to convert them to R code. Just needed to fiddle with customizing the parameters then. Insanely easy really.

I agree it is so much quicker than spinning up a Selenium server and it's implementation for this package would surely reduce dependencies and platform specific errors. Love the properly formatted output too!

Did a quick trial with a giant area request and got a results.json that was essentially empty so don't think it circumvents the tile limitations as such.

Haven't looked at the package source code in detail but breaking up large single requests into multiple smaller requests could be handled via a multi-threaded approach (future_map() etc.) if it isn't already.

If you add me as a contrib/maintainer I can have a look at some of this I reckon.

Cheers,

**Comment by h-a-graham** [comment]
Sounds great - I've sent you an invite as a collaborator. I've also created an issue here: https://github.com/h-a-graham/EAlidaR/issues/49 . 

I'll try to take a look at this in more detail as soon as I have time but any ideas you have would be most welcome! 

Thanks again!
